example/main.c

```diff
@@ -0,0 +1,230 @@
+    }
+
+    params->e820_entries = table->length;
+
+    e820_table_free(table);
+}
+
+static void set_setup_header(struct boot_params *boot_params)
+{
+    boot_params->hdr.vid_mode = 0xFFFF; // VGA
+    boot_params->hdr.type_of_loader = 0xFF;
+    boot_params->hdr.ramdisk_image = 0x0;
+    boot_params->hdr.ramdisk_size = 0x0;
+    boot_params->hdr.ext_loader_ver = 0x0;
+    boot_params->hdr.ext_loader_type = 0x0;
+
+    boot_params->hdr.loadflags |= CAN_USE_HEAP | LOADED_HIGH | KEEP_SEGMENTS;
+    boot_params->hdr.heap_end_ptr =
+        0xe000 - 0x200; // We assume protocol version >= 0x202
+    boot_params->hdr.cmd_line_ptr =
+        REAL_MODE_ADDR + boot_params->hdr.heap_end_ptr;
+}
+
+static void load_image_into_memory(vm_t *vm,
+                                   struct boot_params *boot_params,
+                                   char *image,
+                                   size_t image_size)
+{
+    char *kernel = memory_get_ptr(vm, KERNEL_ADDR);
+
+    if (kernel == NULL)
+    {
+        errx(1, "Failed to get kernel ptr from VM");
+    }
+
+    size_t setup_sectors = boot_params->hdr.setup_sects;
+
+    if (setup_sectors < 4)
+    {
+        setup_sectors = 4;
+    }
+
+    size_t setup_size = (setup_sectors + 1) * 512;
+    size_t kernel_size = image_size - setup_size;
+
+    // Copy the kernel
+    memcpy(kernel, image + setup_size, kernel_size);
+}
+
+static void load_linux(vm_t *vm, char *image, size_t image_size)
+{
+    struct boot_params *boot_params = memory_get_ptr(vm, BOOT_PARAMS_ADDR);
+
+    if (boot_params == NULL)
+    {
+        errx(1, "Failed to get boot_params ptr from VM");
+    }
+
+    memset(boot_params, 0, sizeof(struct boot_params));
+    memcpy(boot_params, image, sizeof(struct boot_params));
+
+    set_setup_header(boot_params);
+    setup_e820(vm, boot_params);
+
+    // Now let's copy the cmd line into the vm memory
+    char *cmdline = memory_get_ptr(vm, boot_params->hdr.cmd_line_ptr);
+
+    if (cmdline == NULL)
+    {
+        errx(1, "Failed to get cmdline ptr from VM");
+    }
+
+    memset(cmdline, 0, boot_params->hdr.cmdline_size);
+    memcpy(cmdline, CMD_LINE, strlen(CMD_LINE));
+
+    load_image_into_memory(vm, boot_params, image, image_size);
+
+    struct kvm_regs regs;
+
+    if (vm_get_regs(vm, &regs) != 1)
+    {
+        errx(1, "Failed to get regs");
+    }
+
+    regs.rsi = BOOT_PARAMS_ADDR;
+    regs.rip = boot_params->hdr.code32_start;
+
+    if (vm_set_regs(vm, &regs) != 1)
+    {
+        errx(1, "Failed to set regs");
+    }
+}
+
+static void *serial_thread(void *param)
+{
+    printf("Thread started\n");
+    serial_t *serial = (serial_t *)param;
+    unsigned char buffer[1024];
+
+    for (;;)
+    {
+        int r = (int)serial_read(serial, buffer, 1024);
+
+        printf("%.*s", r, buffer);
+    }
+
+    return NULL;
+}
+
+int main(int argc, char *argv[])
+{
+    if (argc != 2)
+    {
+        errx(1, "invalid args: ./linux_example <bzimage>");
```

Comment:
Can you please document in README how to compile and start this example, and overall describe how to get a linux image this example can run ? Saying this as it was a request from Pierre Parrend.

Reply:
Done 345de25